File: core/x/javaapi/api/src/main/java/viaduct/java/api/annotations/GraphQLFragment.java
package viaduct.java.api.annotations;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Declares a reusable named GraphQL fragment on a {@link
 * viaduct.java.api.documents.FragmentFromAnnotation} class.
 *
 * <p>The fragment is extracted and validated against the tenant schema at build time. It may be
 * spread from resolver required selections and {@code @GraphQLOperation} documents in the same
 * tenant module.
 */
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
public @interface GraphQLFragment {
  /** A document containing exactly one named GraphQL fragment definition. */
  String value();
}

File: core/tenant/api/src/main/kotlin/viaduct/api/documents/GraphQLFragment.kt
package viaduct.api.documents

import org.intellij.lang.annotations.Language
import viaduct.apiannotations.ExperimentalApi

/**
 * Annotation that declares a named GraphQL fragment on a Kotlin singleton object.
 *
 * Apply this annotation to a Kotlin `object` that extends [FragmentFromAnnotation] to define a
 * reusable GraphQL fragment and spread it (`...MyFragment`) in resolver fragments and
 * `ctx.query`/`ctx.mutation` calls within the same tenant module.
 *
 * ```kotlin
 * @GraphQLFragment("""
 *   fragment UserCoreFields on User {
 *     id
 *     name
 *     email
 *   }
 * """)
 * object UserCoreFieldsFragment : FragmentFromAnnotation<User>()
 * ```
 *
 * The annotated class must be a Kotlin singleton `object`. The fragment text must contain exactly
 * one fragment definition. Named fragments are scoped to the tenant module in which they are
 * declared — spreading a fragment from a different module is a compile error.
 *
 * @see FragmentFromAnnotation
 * @see [Resolver][viaduct.api.resolver.Resolver]
 */
@ExperimentalApi
@Target(AnnotationTarget.CLASS)
@Retention(AnnotationRetention.RUNTIME)
annotation class GraphQLFragment(
    /**
     * The GraphQL fragment definition text. Must contain exactly one fragment definition.
     *
     * Example: `"fragment UserFields on User { id name email }"`
     */
    @Language("GraphQL") val value: String,
)
